workflow/session.go

```diff
@@ -0,0 +1,112 @@
+// within a failed session will return an [TODO] error immediately without even scheduling the activity.
+//
+// If user wants to end a session since some activity returns an error, use the CompleteSession API below,
+// and new session can be created if necessary to retry the whole session.
+//
+// Example:
+//    sessionCtx, err := CreateSession(ctx)
+//    if err != nil {
+//		    // Creation failed. Wrong ctx or too many outstanding sessions.
+//    }
+//    err = ExecuteActivity(sessionCtx, someActivityFunc, activityInput).Get(sessionCtx, nil)
+//    if err != nil {
+//        // Session has failed or activity itself failed.
+//    }
+//    ... // execute more activities using sessionCtx
+//    err = CompleteSession(sessionCtx)
+//    if err != nil {
+//        // Wrong ctx is used or failed to release session resource.
+//    }
+func CreateSession(ctx Context) (Context, error) {
+	return internal.CreateSession(ctx)
+}
+
+// RecreateSession recreate a session based on the sessionInfo passed in. Activities executed within
+// the recreated session will be executed by the same worker as the previous session. RecreateSession()
+// returns an error under the same situation as CreateSession() and has the same usage as CreateSession().
+// It will not check the state of the session described by the sessionInfo passed in, so user can recreate
+// a session based on a failed or completed session.
+//
+// The main usage of RecreateSession is for long sessions that are splited into multiple runs. At the end of
+// one run, complete the current session, get sessionInfo from the context and pass the information to the
+// next run. In the new run, the session can be recreated on the information.
+func RecreateSession(ctx Context, sessionInfo *SessionInfo) (Context, error) {
+	return internal.RecreateSession(ctx, sessionInfo)
+}
+
+// CompleteSession completes a session. It releases worker resources, so other sessions can be created.
+//
+// After a session is completed, user can continue to use the context, but the activities will be scheduled
+// on the normal taskList (as user specified in ActivityOptions) and may be picked up by another worker since
+// it's not in a session.
+//
+// This API will return an error if the session in the context has already completed, or the resource release
+// process has failed. No error will returned if user tries to complete a session that has already failed and
+// the API call won't do anything.
+func CompleteSession(ctx Context) error {
+	return internal.CompleteSession(ctx)
+}
+
+// GetSessionInfo returns the sessionInfo stored in the context. If there are multiple sessions in the context,
+// (for example, the same context is used to create, complete, create another session. Then user found that the
+// session has failed, and created a new one on it), the most recent sessionInfo will be returned.
+//
+// This API will return an error if there's no sessionInfo in the context.
```

Comment:
I would make the API return pointer and no error.

Reply:
Done.